service/internal/fanoutconsumer/logs.go

```diff
@@ -80,3 +83,50 @@ func (lsc *logsConsumer) ConsumeLogs(ctx context.Context, ld plog.Logs) error {
 	}
 	return errs
 }
+
+var _ connector.LogsRouter = (*logsRouter)(nil)
+
+type logsRouter struct {
+	consumer.Logs
+	consumers map[component.ID]consumer.Logs
+}
+
+func NewLogsRouter(cm map[component.ID]consumer.Logs) consumer.Logs {
+	consumers := make([]consumer.Logs, 0, len(cm))
+	for _, consumer := range cm {
+		consumers = append(consumers, consumer)
+	}
+	return &logsRouter{
+		Logs:      NewLogs(consumers),
+		consumers: cm,
+	}
+}
+
+func (r *logsRouter) PipelineIDs() []component.ID {
+	ids := make([]component.ID, 0, len(r.consumers))
+	for id := range r.consumers {
+		ids = append(ids, id)
+	}
+	return ids
+}
+
+func (r *logsRouter) Consumer(pipelineIDs ...component.ID) (consumer.Logs, error) {
+	if len(pipelineIDs) == 0 {
+		return nil, fmt.Errorf("missing consumers")
+	}
+	consumers := make([]consumer.Logs, 0, len(pipelineIDs))
+	var errors error
+	for _, pipelineID := range pipelineIDs {
+		c, ok := r.consumers[pipelineID]
+		if ok {
+			consumers = append(consumers, c)
+		} else {
+			errors = multierr.Append(errors, fmt.Errorf("missing consumer: %q", pipelineID))
+		}
+	}
```

Comment:
I would go with simpler solution and just return at first error. Since this is not a "user input" showing just the first error is enough.

Reply:
Is it ok if we leave it so that users can see all errors?

Reply:
I think "it is ok". But the question I have, is it worth it? Does it provide any benefit? It is hard to imagine how you will get to this, since the components that use this API have access to all "piepeline IDs", so this to me seem like a bug in that component not a user error.

Reply:
Here is an example of when a user would see this error:

```yaml
receivers:
  foo:
exporters:
  bar/left:
  bar/right:
connectors:
  routing:
    table:
      - statement: ... 
        pipelines: [logs/left]
      - statement: ...
        pipelines: [logs/right, logs/fake, logs/oops]
service:
  pipelines:
    logs/in:
      receivers: [foo]
      exporters: [routing]
    logs/left:
      receivers: [routing]
      exporters: [bar/left]
    logs/right:
      receivers: [routing]
      exporters: [bar/right]
```

Routing connector constructor will ask for consumers and this will detect invalid user config when validating the second route. The error informs the user which of the specified pipelines are invalid.
```go
func createLogs(_ context.Context, _ connector.CreateSettings, cfg component.Config, logs consumer.Logs) (connector.Logs, error) {
	c := cfg.(RoutingConnectorConfig)
	logsRouter := logs.(connector.LogsRouter)
	
	// generate consumer for each route only once
	routeConsumers := make([]consumer.Logs, 0, len(c.Table))
	for i, route := range c.Table {
		cons, err := logsRouter.Consumer(route.Pipelines...)
		if err != nil {
			return nil, err // missing consumer: "logs/fake", missing consumer: "logs/oops"
		}
		routeConsumers[i] = cons 
	}
	...
	return &logsRoutingConnector{
		routeConsumers: routeConsumers,
		...
	}
}

```

Reply:
If you really want that you can check the IDs since you have access to all of them and error across all "tables". Still not convinced it is worth it.